File: api/src/main/scala/com/variant/demo/bookworms/Variant.scala
package com.variant.demo.bookworms

import akka.http.scaladsl.model.HttpResponse
import akka.http.scaladsl.model.headers.Referer
import akka.http.scaladsl.server.RequestContext
import com.typesafe.config.ConfigFactory
import com.typesafe.scalalogging.LazyLogging
import com.variant.client.stdlib.SessionIdTrackerAkka
import com.variant.client._
import com.variant.share.schema.State

import java.util.Optional
import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._
import scala.util.{Failure, Success, Try}

/**
 * General purpose helper functions.
 */
object Variant extends LazyLogging {

  private val config = ConfigFactory.load()
  private val uri = config.getString("bookworms.variant.uri")
  private val client: VariantClient = VariantClient.build {
    builder => builder.withSessionIdTrackerClass(classOf[SessionIdTrackerAkka])
  }
  private var _connection: Option[Connection] = None

  // The connection to Variant server.
  private def connection(): Option[Connection] = {
    if (_connection.isEmpty) _connection =
      try {
        logger.info("(Re)connecting to Variant schema [" + uri + "]")
        val result = client.connectTo(uri);
        logger.info("Connected to Variant URI [" + uri + "]")
        Some(result)
      }
      catch {
        case sce: ServerConnectException =>
          logger.error(sce.getMessage)
          None
        case t: Throwable =>
          logger.error("Failed to connect to Variant URI [" + uri + "]", t)
          None
      }
    _connection
  }

  /** Get existing session tracked by the session ID cookie */
  def thisVariantSession(implicit ctx: RequestContext): Option[Session] = {
    connection().flatMap(_.getSession(ctx.request).toScala)
  }

  /** Infer the Variant state from the referring page. */
  private def inferState(ssn: Session)(implicit ctx: RequestContext): Option[State] = {
    val refererUri: String = Referer.parseFromValueString(ctx.request.getHeader("Referer").get().value()) match {
      case Right(refererHeader) => refererHeader.uri.path.toString
      case Left(_) => throw new Exception("No Referer Header")
    }
    ssn.getSchema.getStates.asScala.find(state => refererUri.matches(state.getParameters.get("path")))
  }

  /** Target current session for a given state */
  def targetForState()(implicit ctx: RequestContext): Option[StateRequest] = {
    try {
      for {
        ssn <- connection().map(
          // Avoid concurrent state creation if does not exist. This is only an issue on the very first page,
          // where we don't yet have a variant session.
          conn => this.synchronized {
            conn.getOrCreateSession(ctx.request, Optional.of(UserRegistry.currentUser))
          }
        )
        myState <- inferState(ssn)
      }
      yield {
        ssn.setAttribute("isInactive", UserRegistry.isInactive.toString)
        ssn.targetForState(myState)
      }
    } catch {
      case _: ServerConnectException => None
    }
  }

  /** Is a given experience live in a given state request */
  def isExperienceLive(stateRequest: StateRequest, variationName: String, experienceName: String): Boolean = {
    stateRequest.getLiveExperience(variationName).stream().anyMatch(exp => exp.getName == experienceName)
  }

  def commitOrFail(req: StateRequest): Try[HttpResponse] => Try[HttpResponse] = {
    case Success(goodHttpResponse) =>
      val newHttpResp = req.commit(goodHttpResponse).asInstanceOf[HttpResponse]
      Try(newHttpResp)
    case Failure(ex) =>
      req.fail()
      Failure(ex)
  }
}

File: servlet-spring-java/webapp/src/main/java/com/variant/demo/Variant.java
package com.variant.demo;

import com.variant.client.*;
import com.variant.client.stdlib.SessionIdTrackerServlet5;
import com.variant.share.schema.State;
import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Optional;
import java.util.Properties;

public class Variant {

	final private static Logger logger = LoggerFactory.getLogger(Variant.class);

	final private static String url;
	static {
		Properties props = new Properties();
		try {
			props.load(ClassLoader.getSystemResourceAsStream("application.properties"));
			url = props.getProperty("variant.server.url");
		}
		catch (IOException e) {
			throw new RuntimeException("Unable to load application properties: " + e.getMessage(), e);
		}
	}

	final private static VariantClient client = VariantClient
		.build(builder -> builder.withSessionIdTrackerClass(SessionIdTrackerServlet5.class));

	private static Optional<Connection> _connection = Optional.empty();

	// The connection to Variant server.
	private static Optional<Connection> connection() {
		if (_connection.isEmpty()) {
			try {
				logger.info("(Re)connecting to Variant schema [" + url + "]");
				Connection result = client.connectTo(url);
				logger.info("Connected to Variant URI [" + url + "]");
				_connection = Optional.of(result);
			}
			catch (ServerConnectException sce) {
				logger.error(sce.getMessage());
				_connection = Optional.empty();
			}
			catch (Throwable t) {
				logger.error("Failed to connect to Variant URI [" + url + "]", t);
				_connection = Optional.empty();
			}
		}
		return _connection;
	}

	/** Infer the Variant state from the referring page's URL. */
	public static Optional<State> inferState(Session ssn) {

		HttpServletRequest httpServletRequest = ((ServletRequestAttributes) RequestContextHolder
			.currentRequestAttributes()).getRequest();

		return Optional.ofNullable(httpServletRequest.getHeader("referer")).flatMap(refererString -> {
			try {
				String path = new URL(refererString).getPath();
				return ssn.getSchema()
					.getStates()
					.stream()
					.filter(state -> path.matches(state.getParameters().get("path")))
					.findAny();
			}
			catch (MalformedURLException ex) {
				throw new RuntimeException(ex);
			}
		});
	}

	/** Get Variant session. */
	public static Optional<Session> getVariantSession() {
		try {
			HttpServletRequest httpServletRequest = ((ServletRequestAttributes) RequestContextHolder
				.currentRequestAttributes()).getRequest();
			return connection().map(conn -> conn.getOrCreateSession(httpServletRequest,
					Optional.of(VariantController.loggedInOwner.owner.getFullName())));
		}
		catch (ServerConnectException scx) {
			// The server is down. We'll attempt to reconnect.
			return Optional.empty();
		}
	}

	/** Is a given experience live in a given state request */
	public static Boolean isExperienceLive(StateRequest stateRequest, String experimentName, String experienceName) {
		return stateRequest.getLiveExperience(experimentName)
			.map(exp -> exp.getName().equals(experienceName))
			.orElse(false);
	}

}
